tests/mock_network/churn.rs

```diff
@@ -219,71 +456,91 @@ fn check_added_indices(
+
+    let dropped_names = if section_num > min_section_num {
+        drop_random_nodes(rng, nodes, drop_probability)
+    } else {
+        BTreeSet::new()
+    };
+
+    let added_indices = if section_num < max_section_num {
+        add_nodes(rng, &network, nodes, add_probability)
+    } else {
+        BTreeSet::new()
+    };
+
+    (added_indices, dropped_names)
+}
 
-    // Drop failed_indices and poll remaining nodes to clear pending states.
-    for index in failed_indices.into_iter().rev() {
-        drop(nodes.remove(index));
-    }
+fn progress_and_verify<R: Rng>(
+    rng: &mut R,
+    network: &Network,
+    nodes: &mut [TestNode],
+    message_schedule: MessageSchedule,
+    added_indices: BTreeSet<usize>,
+    dropped_names: BTreeSet<XorName>,
+) {
+    let expectations = match message_schedule {
+        MessageSchedule::AfterChurn => {
+            poll_after_churn(nodes, added_indices, dropped_names);
+            let expectations = setup_expectations(rng, nodes, network.elder_size());
+            poll_and_resend(nodes);
+            expectations
+        }
+        MessageSchedule::DuringChurn => {
+            let expectations = setup_expectations(rng, nodes, network.elder_size());
+            poll_after_churn(nodes, added_indices, dropped_names);
+            expectations
+        }
+    };
 
-    poll_and_resend(&mut nodes);
+    expectations.verify(nodes);
+    verify_invariant_for_all_nodes(network, nodes);
     shuffle_nodes(rng, nodes);
-
-    added_names
 }
 
-// Churns the given network randomly. Returns any newly added indices.
-// If introducing churn, would either drop/add nodes in each prefix.
-fn random_churn<R: Rng>(
-    rng: &mut R,
-    network: &Network,
-    nodes: &mut Vec<TestNode>,
-    max_prefixes_len: usize,
-) -> BTreeSet<usize> {
-    // 20% chance to not churn.
-    if rng.gen_weighted_bool(5) {
-        return BTreeSet::new();
-    }
+fn poll_after_churn(
+    nodes: &mut [TestNode],
+    added_indices: BTreeSet<usize>,
+    dropped_names: BTreeSet<XorName>,
+) {
+    trace!(
+        "Adding {{{:?}}}, dropping {:?}",
+        added_indices
+            .iter()
+            .map(|index| nodes[*index].name())
+            .format(", "),
+        dropped_names
+    );
 
-    let section_count = count_sections(nodes);
-    if section_count < max_prefixes_len {
-        return add_nodes(rng, &network, nodes, false);
-    }
+    poll_and_resend(nodes);
+    let added_names = check_added_indices(nodes, added_indices);
 
-    // Use elder_size rather than section size to prevent collapsing any groups.
-    let max_drop = (unwrap!(nodes[0].inner.elder_size()) - 1)
-        * (QUORUM_DENOMINATOR - QUORUM_NUMERATOR)
-        / QUORUM_DENOMINATOR;
-    assert!(max_drop > 0);
-    let dropped_nodes = drop_random_nodes(rng, nodes, Some(max_drop));
-    warn!("Dropping nodes: {:?}", dropped_nodes);
-    BTreeSet::new()
+    warn!("Added {:?}, dropped {:?}", added_names, dropped_names);
```

Comment:
there are benefit in putting these in higher log level as well since they are useful to understand test tests even if we want to only get warning or errors.